include/pybind11/detail/common.h

```diff
@@ -324,7 +324,8 @@ PYBIND11_WARNING_POP
 #endif
 
 // See description of PR #4246:
-#if !defined(NDEBUG) && !defined(PY_ASSERT_GIL_HELD_INCREF_DECREF) && !defined(PYPY_VERSION)      \
+#if !defined(PYBIND11_NO_GIL_CHECKS) && !defined(NDEBUG)                                          \
+    && !defined(PY_ASSERT_GIL_HELD_INCREF_DECREF) && !defined(PYPY_VERSION)                       \
```

Comment:
I agree this is what we need to do.

We don't need `PY_ASSERT_GIL_HELD_INCREF_DECREF` in that case, I can use the new name instead for the purpose this was meant for.

So this could simply become:

```
#if !defined(PYBIND11_NO_ASSERT_GIL_HELD_INCREF_DECREF) && !defined(NDEBUG) && !defined(PYPY_VERSION) \
```

I think it's important to make the names as similar as possible. We have another GIL check already that's not affected. Possibly we have more later that we also don't want to turn off with the new define here.

Reply:
> We don't need PY_ASSERT_GIL_HELD_INCREF_DECREF in that case

Are you sure? Your comments in the original PR made it sound like PY_ASSERT_GIL_HELD_INCREF_DECREF would still be necessary here as it's a CPython thing.

> PYBIND11_NO_ASSERT_GIL_HELD_INCREF_DECREF

Swapped to the new name. Good point about making things more clear.

Reply:
> it's a CPython thing

That was my idea at the time, but I haven't been able to actually squash all non-pybind11 failures (only ~3 I'm aware of). I don't think I'll have the free energy to see this through anytime soon, and convince core Python developers to adopt it. So let's not make this a thing here, the list of conditions is already easily confusing.

Reply:
Ok. Removed.